support bond type gt4 in MDLV3000Reader

uli-f opened the pull request:
Most of my questions are related to how query objects (atoms, bond, container) are implemented and supposed to be used in CDK.

Few comments/questions:
- Re commit [f98fa84](https://github.com/cdk/cdk/commit/f98fa84c98efb47e17b5314b10c58ee5e80a7670): I assume that the four methods and the class attribute ended up being non-private by accident. The argument `MDLV3000Reader.ReadState` common to the four methods is a private inner static class so calling any of these methods from outside `MDLV2000Reader` should not have been possible in the first place. Please feel free to revert if this is too aggressive a change, that's why it's done in a separate commit.
After the commit I noticed there are two more methods that probably shouldn't be public: `MDLV3000Reader::exaustStringTokenizer` and `MDLV3000Reader::readLine`. Personally, I would also make them private.
- I don't fully understand why the actions related to query atoms / bonds are currently implemented as they are in `MDLV2000Reader`. Whether the currently processed atom container has query features is determined on-the-fly in `MDLV2000Reader`. It seems to me that most/all these operations might be better done as a post-op in the reader once reading of an atom container has finished and before returning it (i.e., at the beginning of `MDLV3000Reader::finalizeMol`). But it might also just miss sth here 🤷🏼‍♂️ 
- I stumbled (again?) upon the issue of not being able to get `real` bonds from the `AtomContainer` but end up with `BaseBondRef` and `QueryBondRef`, respectively. This limits the usefulness of the tests related to query bond features I added in `MDLV3000Reader`. I'd appreciate any advice of how to best get an actual `(I)QueryBond` so that I can test the expression that's associated with the query bond.
- Looking at MDLV2000Reader I cannot figure out what the criterion is for an `AtomContainer` to get altered to a `QueryAtomContainer`.
  - Is it the presence of at most one QueryBond (this is what I see in code)?
  - Is it the presence of at most one QueryAtom (I don't see this in code)?
- I did not consider anything related to **query atoms**. It might be worthwhile to alter any atom objects to query atoms if we are dealing with an atom container with at least one query bond. At least, this is what I gathered from `MDLV2000Reader`. This might be implemented as a processing step in `MDLV3000Reader::finalizeMol`.
- I thought it worthwhile to clarify my questions in this PR and then do a separate PR for the `MDLV3000Writer` that might benefit from my better insight.

And a slightly unrelated side notes: I also miss language features from 17 and 21. Any plans of stepping up the minimum requirement to at least 17?

uli-f commented:
partly address #1061

johnmay commented:
1. Private looks fine there, I think the intention was they may be reused by a reaction reader at some point, note non-private != public and were package-private which is generally quite useful. But since there is no break it's all OK
2. The whole query API sort of needs a rethink, I actually would prefer to just store these on a normal IAtomContainer and set a flag "is query". Either way
3. You can ``BondDeref.deref(bond)`` it and then cast it, ideally I'd like IQueryAtom to expose the Expression at the interface level but needs some thought.  For the tests why not just do the test by checking a match... i.e. a single_or_aromatic should match a single, aromatic and not a double or triple. That also guards against future internal changes since providing it still does what you want it doesn't mater about the internal state.
4. I believe it's both, however it's common for people to misuse aromatic bonds (which are a query feature) to represent real molecules.


Could you improve the tests as per (3) checking as ((IQueryBond)bond).matches(expected) and I will merge.

Re Java 17/21 upgrade will be CDK 3.0 - we plan to start that branch at some point.

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -157,10 +145,21 @@ public <T extends IChemObject> T read(T object) throws CDKException {
-        IAtomContainer mol;
-        int dimensions = 0; // 0D (undef/no coordinates), 2D, 3D
-        boolean chiral;
+        IAtomContainer atomContainer;
```
Generally I prefer mol for it's terseness, atomContainer is such a bad name :-). There was an IMolecule which didn nothing and was removed in CDK 1.4

uli-f replied:
Changed accordingly :)

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -239,26 +251,45 @@ public IAtomContainer readConnectionTable(IChemObjectBuilder builder) throws CDK
+        // migrate atom container to IQueryAtomContainer implementation if there are any objects with query features
+        if (state.isQuery) {
+            // shallow copy of the original atom container, i.e. same atoms and electron containers as original
+            final IQueryAtomContainer queryAtomContainer = new QueryAtomContainer(readAtomContainer, readAtomContainer.getBuilder());
```
This isn't quite right, it will work but I would "deref" the atoms and bonds from the initial since IQueryAtomContainer isn't implemented in terms of that. 

Ideally in future IQueryAtomContainer will go away or at least have better internals. Right now it's not a priority as performance of query manipulation isn't that critical.

uli-f replied:
Not quite sure what you mean here 🤷🏼

johnmay replied:
So the current non-query AtomContainer's wrap every atom and bond in a ref. This is basically a hack to allow Atoms and Bonds to be added to a multiple atom containers yet when you access an atom/bond from a container it knows where it came from. Hence you can ask things like "atom.bonds()" and it works.

The QueryAtomContainer is "dump" and doesn't provide this semantics. 
Option 1) Unbox/deref the atoms/bonds before they go in (quick)
Option 2) Make the QueryAtomContainer smarter (best)

I'll do option 1 by a separate PR.

johnmay replied:
If I was designing the API from scratch I would not allow an atom/bond to be in multiple Molecules at once, it's not as much of a benefit as allowing each atom/bond to be "owned" by a single mol.

johnmay replied:
Fixed it now in #1105

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -267,18 +298,29 @@ private void finalizeMol(ReadState state) {
-        if (!isQuery)
-            finalizeStereochemistry(state, readData);
+        // skip infering and setting stereo chemistry if atom container has query features or aromatic bond
+        if (!isQueryOrAromaticBond)
```
Not for this PR but we should tweak the stereo at some point, I think the reason it is skipped is the stereo assignment looks for 4 single bonds for tetrahedral (for example)

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -366,29 +413,33 @@ private void finalizeDimensions(ReadState state) {
         char ch;
 
-        while (i < len && isDigit(ch = str.charAt(i))) {
+        while (i < len && Character.isDigit(ch = str.charAt(i))) {
```
No please use isDigit

johnmay replied:
MOLfile's are ASCII and that isn't Character.isDigit checks.

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -406,15 +457,15 @@ private void parseStereoGroup(Map<Integer,Integer> flags, String str, int type)
 
         // skip the count since we're storing in map
-        while (i < len && isDigit(str.charAt(i)))
+        while (i < len && Character.isDigit(str.charAt(i)))
```
ditto

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -406,15 +457,15 @@ private void parseStereoGroup(Map<Integer,Integer> flags, String str, int type)
         while (i < len) {
             int val = 0;
-            while (i < len && isDigit(ch = str.charAt(i))) {
+            while (i < len && Character.isDigit(ch = str.charAt(i))) {
```
ditto

johnmay commented on storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:
```diff
@@ -796,35 +900,42 @@ public void readBondBlock(ReadState state) throws CDKException {
 
-                logger.debug("Added bond: " + bond);
+                // set flags of atoms participating in bond to aromatic if bond has flag set to aromatic
+                if (bond.getFlag(IChemObject.AROMATIC)) {
```
bond.isAromatic()

johnmay commented on storage/ctab/src/test/java/org/openscience/cdk/io/MDLV3000ReaderTest.java:
```diff
@@ -234,20 +226,136 @@ void hydIsotopesAsPseudo() throws Exception {
+            // bond assertions
+            assertThat(atomContainer.getBondCount(), is(2));
+            // actual implementing class is AtomContainer.BaseBondRef
+            assertThat(atomContainer.getBond(0).getClass().getName().contains("Query"), is(false));
```
assert atomContainer.getBond(0) instanceof IQueryAtom
assert !atomContainer.getBond(0).matches(single_alip)
assert !atomContainer.getBond(0).matches(double_alip)
assert atomContainer.getBond(0).matches(double_arom)
assert atomContainer.getBond(0).matches(single_arom)

johnmay reviewed (commented):
Few minor things

johnmay commented:
@uli-f Are you ok to make the updates?

uli-f commented:
> @uli-f Are you ok to make the updates?

Yes.
And thank you for your valuable feedback on this.
Just came back from a two week vacation, will get right back on this now!

uli-f commented:
> @uli-f Are you ok to make the updates?

Yes.
And thank you for your valuable feedback on this.
Just came back from a two week vacation, will get right back on this now!

uli-f commented:
> 1. Private looks fine there, I think the intention was they may be reused by a reaction reader at some point, note non-private != public and were package-private which is generally quite useful. But since there is no break it's all OK

I changed `MDLV3000Reader::exaustStringTokenizer` and `MDLV3000Reader::readLine` to private.

>     2. The whole query API sort of needs a rethink, I actually would prefer to just store these on a normal IAtomContainer and set a flag "is query". Either way

I left as is for now.

But very happy to just set a flag in a normal `IAtomContainer` instead of converting to a `QueryAtomContainer`. If you give me more details about your preference (add `IAtomContainer::isQuery`? Or just a `CDKConstant` set as a property on `IAtomContainer`?) I can change this PR accordingly.
 
>     3. You can `BondDeref.deref(bond)` it and then cast it, ideally I'd like IQueryAtom to expose the Expression at the interface level but needs some thought.  For the tests why not just do the test by checking a match... i.e. a single_or_aromatic should match a single, aromatic and not a double or triple. That also guards against future internal changes since providing it still does what you want it doesn't mater about the internal state.

Good idea to check a match for the query bonds in tests.

>     4. I believe it's both, however it's common for people to misuse aromatic bonds (which are a query feature) to represent real molecules.

Just want to emphasize again that anything related to **query atoms** is not considered in this PR. I added a paragraph to the javadocs of the class.

> Could you improve the tests as per (3) checking as ((IQueryBond)bond).matches(expected) and I will merge.
> 
> Re Java 17/21 upgrade will be CDK 3.0 - we plan to start that branch at some point.
